File: backend/today/github.py
import re
import requests
from bs4 import BeautifulSoup


def clean_string(string: str):
    cleaned = string.strip()

    cleaned = re.sub(r"\s+", " ", cleaned)
    cleaned = re.sub(" ", "", cleaned)
    cleaned = re.sub(r"[^a-zA-Z0-9,//\s]", "", cleaned)

    return cleaned


async def get_github_trending():
    response = []
    html = requests.get("https://github.com/trending").text
    soup = BeautifulSoup(html, "html.parser")
    rows = soup.find_all("article")

    for row in rows:
        name_block = row.find("h2").find("a")
        url = f"https://github.com{name_block['href']}"

        repo_split = name_block.get_text(strip=True).replace(" /", " / ").split(" / ")
        repo_owner = repo_split[0]
        repo_name = repo_split[1]

        description_element = row.find("p")
        description = (
            description_element.get_text(strip=True) if description_element else ""
        )
        language_raw = row.find("span", itemprop="programmingLanguage")
        language = language_raw.text if language_raw else None
        stars = clean_string(row.find("svg", attrs={"aria-label": "star"}).parent.text)

        response.append(
            {
                "repo_name": repo_name,
                "repo_owner": repo_owner,
                "description": description,
                "language": language,
                "stars": stars,
                "url": url,
            }
        )

    return response

File: lib/sources/github.ts
import * as cheerio from "cheerio";

export type GitHubRepo = {
  repoName: string;
  repoOwner: string;
  description: string;
  language: string | null;
  stars: string;
  url: string;
};

export const GITHUB_TAG = "github";

/**
 * Trim, collapse whitespace, strip spaces entirely, then drop anything that is
 * not alphanumeric, a comma or a slash. Turns "\n  1,234  \n" into "1,234".
 */
const cleanStars = (value: string) =>
  value
    .trim()
    .replace(/\s+/g, "")
    .replace(/[^a-zA-Z0-9,/]/g, "");

export const getGitHubTrending = async (): Promise<GitHubRepo[]> => {
  const response = await fetch("https://github.com/trending", {
    next: { revalidate: 3600, tags: [GITHUB_TAG] },
  });

  if (!response.ok) {
    throw new Error(`GitHub Trending returned ${response.status}`);
  }

  const $ = cheerio.load(await response.text());

  return $("article")
    .toArray()
    .flatMap((article) => {
      const element = $(article);
      const href = element.find("h2 a").attr("href");

      // The href is "/owner/repo", which is far less brittle than parsing the
      // anchor's text and splitting on " / ".
      const [, repoOwner, repoName] = (href ?? "").split("/");

      if (!repoOwner || !repoName) {
        return [];
      }

      const language = element
        .find('[itemprop="programmingLanguage"]')
        .first()
        .text()
        .trim();

      return [
        {
          repoName,
          repoOwner,
          description: element.find("p").first().text().trim(),
          language: language || null,
          stars: cleanStars(
            element.find('svg[aria-label="star"]').first().parent().text(),
          ),
          url: `https://github.com${href}`,
        },
      ];
    });
};
